### PR TITLE
improve mouse interaction

### PR DESCRIPTION
**Pull request by gimler**
* add `clickWithLeftButton`
* make element optional for `clickWithLeftButton`, `clickWithRightButton` and `moveMouseOver`
* improve documentation

### CODE REVIEW

**Comment by DavertMik** [comment]
How does `clickWithLeftButton` differs from `click`?

**Comment by gimler** [comment]
Click use findclickable to find a clickable element and click on it.

Clickwithleftbutton click with left button on coordinates on page or any element.

**Comment by DavertMik** [comment]
> Clickwithleftbutton click with left button on coordinates on page or any element.

anyway, the title and the reference for this method completely misleading. Could you provide some use cases for that? Also this may require additional parameter like coordintates to be passed. If the overall idea is to click by coords they should be available to set

**Comment by gimler** [comment]
We have a svg seating plan with 300 seats and we must click the seats to book. I have make the function like the right click function

We can add a check that you must provide min element or coordinates or both.

**Comment by gimler** [comment]
We try to switch from behat to codeception an behat has the functionality.

**Comment by DavertMik** [comment]
Ok, I agree that this functionality should be added I just want to ensure that the API is clear and easy to use. 

How about `clickByCoordinates($rootElement, $offset)`?

**Comment by DavertMik** [comment]
Or maybe better:

`clickElement($locator, $offsetX, $offsetY)`
 
probably this offset should also be added to `clickWithRightButton`

**Comment by gimler** [comment]
Agree the x and y parameter make sense

Your proposal
```
clickElement($locator = null, $offsetX = null, $offsetY = null)
clickWithRightButton($cssOrXPath = null, $offsetX = null, $offsetY = null)
```

My proposal similar name for left and right mouse click
```
clickWithLeftButton($rootElement = null, $offsetX = null, $offsetY = null)
clickWithRightButton($rootElement = null, $offsetX = null, $offsetY = null)
```

see also the function from facebook webdriver that is called `Facebook\WebDriver\Remote\RemoteMouse::mouseMove`

```
    /**
     * @param WebDriverCoordinates $where
     * @param int|null $x_offset
     * @param int|null $y_offset
     *
     * @return RemoteMouse
     */
    public function mouseMove(
        WebDriverCoordinates $where = null,
        $x_offset = null,
        $y_offset = null
    ) {
        $params = [];
        if ($where !== null) {
            $params['element'] = $where->getAuxiliary();
        }
        if ($x_offset !== null) {
            $params['xoffset'] = $x_offset;
        }
        if ($y_offset !== null) {
            $params['yoffset'] = $y_offset;
        }
        $this->executor->execute(DriverCommand::MOVE_TO, $params);

        return $this;
    }
```

**Comment by edno** [comment]
Why not aligning naming with php-we driver where right click is [contextClick](https://github.com/facebook/php-webdriver/blob/master/lib/interactions/internal/WebDriverContextClickAction.php),?

**Comment by gimler** [comment]
You mean contextClick() instate of clickWithRightButton that would be a unnessasary bc break.

**Comment by DavertMik** [comment]
>  My proposal similar name for left and right mouse click

```
clickWithLeftButton($rootElement = null, $offsetX = null, $offsetY = null)
clickWithRightButton($rootElement = null, $offsetX = null, $offsetY = null)
```

:+1: Agreed! Looks good and consistent. Thanks for pointing that.

> Why not aligning naming with php-webdriver where right click is contextClick,?

Because we already have it named `clickWithRightButton` and I don't see any valuable reason to implement a new naming. We should maintain BC policy.

File: src/Codeception/Module/WebDriver.php [hunk 1]
```diff
@@ -2177,20 +2179,52 @@ public function dragAndDrop($source, $target)
...
+        if (null === $cssOrXPath) {
+            $el = $this->matchFirstOrFail($this->webDriver, $cssOrXPath);
+        }
+
         $this->webDriver->getMouse()->mouseMove($el->getCoordinates(), $offsetX, $offsetY);
     }
 
     /**
-     * Performs contextual click with the right mouse button on an element.
+     * Performs contextual click with the right mouse button on an element or page.
```
Comment (DavertMik) [inline src/Codeception/Module/WebDriver.php]: Ok, please update the description of this method. As you see it says "right mouse button"

File: src/Codeception/Module/WebDriver.php [hunk 2]
```diff
@@ -2177,20 +2179,52 @@ public function dragAndDrop($source, $target)
...
+     *
+     * @param string $cssOrXPath css or xpath of the web element
      *
-     * @param $cssOrXPath
      * @throws \Codeception\Exception\ElementNotFound
      */
-    public function clickWithRightButton($cssOrXPath)
+    public function clickWithLeftButton($cssOrXPath = null)
```
Comment (DavertMik) [inline src/Codeception/Module/WebDriver.php]: update signature to provide mouse coordinates

File: src/Codeception/Module/WebDriver.php [hunk 3]
```diff
@@ -2177,20 +2179,52 @@ public function dragAndDrop($source, $target)
...
+     * $I->clickWithRightButton();
+     * $I->clickWithRightButton(['css' => '.checkout']);
+     * ?>
+     * ```
+     *
+     * @param string $cssOrXPath css or xpath of the web element
+     *
+     * @throws \Codeception\Exception\ElementNotFound
+     */
+    public function clickWithRightButton($cssOrXPath = null)
```
Comment (DavertMik) [inline src/Codeception/Module/WebDriver.php]: update signature to provide mouse coordinates

**Comment by gimler** [comment]
I have implemented all but i did not get the test pass for right click in phantomjs ;(

**Comment by gimler** [comment]
ok the test fail because the latest phantomjs did not has the latest ghostdriver see https://github.com/ariya/phantomjs/issues/14005.

Should i remove or comment the right click test case?

**Comment by Naktibalda** [comment]
@gimler Your tests fail with Selenium too: 
https://semaphoreci.com/codeception/codeception/branches/pull-request-3864/builds/3
```
---------
1) WebDriverTest: Right click
 Test  tests/web/WebDriverTest.php:testRightClick
Failed asserting that  on page /form/click
--> context, offsetX: undefined - offsetY: undefined
--> contains "click, offsetx: 123 - offsety: 88".
#1  /home/runner/Codeception/src/Codeception/Module/WebDriver.php:2506
#2  /home/runner/Codeception/src/Codeception/Module/WebDriver.php:725
#3  /home/runner/Codeception/tests/web/WebDriverTest.php:791
#4  WebDriverTest->testRightClick
#5  /home/runner/Codeception/codecept:34

---------
2) WebDriverTest: Left click
 Test  tests/web/WebDriverTest.php:testLeftClick
Failed asserting that  on page /form/click
--> click, offsetX: 600 - offsetY: 117
--> contains "click, offsetx: 600 - offsety: 384".
#1  /home/runner/Codeception/src/Codeception/Module/WebDriver.php:2506
#2  /home/runner/Codeception/src/Codeception/Module/WebDriver.php:725
#3  /home/runner/Codeception/tests/web/WebDriverTest.php:774
#4  WebDriverTest->testLeftClick
#5  /home/runner/Codeception/codecept:34
```

Please pull before making any more changes,  I merged 2.2 branch to resolve conflicts and make SemaphoreCI run.

**Comment by gimler** [comment]
webdriver is green phantom fails on testRightClick see my comments above. ready for review

**Comment by gimler** [comment]
Build is green please. Review.

**Comment by DavertMik** [comment]
Thanks, I am on vacation, I will be online on monday and will review and merge at that time

**Pull request merged**